Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts

```diff
@@ -0,0 +1,209 @@
+import * as Q from 'q';
+import * as os from 'os';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as tl from 'vsts-task-lib/task';
+
+import * as providers from 'artifact-engine/Providers';
+import { HttpClientResponse } from 'artifact-engine/Providers/typed-rest-client/HttpClient';
+
+export class CommitsDownloader {
+    private webProvider: providers.WebProvider;
+
+    constructor(webProvider: providers.WebProvider) {
+        this.webProvider = webProvider;
+    }
+
+    public DownloadFromSingleBuildAndSave(buildId: string): Q.Promise<string> {
+        let defer: Q.Deferred<string> = Q.defer<string>();
+        
+        console.log(tl.loc("GettingCommitsFromSingleBuild", buildId));
+        this.GetCommitsFromSingleBuild(buildId).then((commits: string) => {
+            this.UploadCommits(commits).then(() => {
+                defer.resolve(commits);
+            }, (error) => {
+                defer.reject(error);
+            });
+        }, (error) => {
+            defer.reject(error);
+        });
+
+        return defer.promise;
+    }
+
+    public DownloadFromBuildRangeAndSave(startBuildId: string, endBuildId: string): Q.Promise<string> {
+        let defer: Q.Deferred<string> = Q.defer<string>();
+
+        console.log(tl.loc("DownloadingCircleCIChangeBetween", startBuildId, endBuildId));
+        this.GetCommits(startBuildId, endBuildId).then((commits: string) => {
+            this.UploadCommits(commits).then(() => {
+                defer.resolve(commits);
+            }, (error) => {
+                defer.reject(error);
+            });
+        }, (error) => {
+            defer.reject(error);
+        });
+
+        return defer.promise;
+    }
+
+    private GetCommitsFromSingleBuild(buildId: any): Q.Promise<string> {
+        let connection = tl.getInput("connection", true);
+        let definitionId = tl.getInput("definition", true);
+        var endpointUrl = tl.getEndpointUrl(connection, false);
+        let defer = Q.defer<string>();
+        let url: string = endpointUrl + "/api/v1.1/project/"+definitionId+"/"+buildId;
```

Comment:
handle double slashes?

Reply:
can you use string formater?
`${endpointUrl}/api/v1.1/project/${definitionId}/${buildId};

Reply:
also chomp of trailing slash in endpoint which is most likely.